pkg/api/rest/rest.go

```diff
@@ -202,20 +202,19 @@ type Redirector interface {
 	ResourceLocation(ctx api.Context, id string) (remoteLocation *url.URL, transport http.RoundTripper, err error)
 }
 
-// ConnectHandler is a handler for HTTP connection requests. It extends the standard
-// http.Handler interface by adding a method that returns an error object if an error
-// occurred during the handling of the request.
-type ConnectHandler interface {
-	http.Handler
-
-	// RequestError returns an error if one occurred during handling of an HTTP request
-	RequestError() error
+// Responder provides standard response behavior.
+type Responder interface {
+	// Object writes the provided object to the response. Invoking this method multiple times is undefined.
+	Object(statusCode int, obj runtime.Object)
+	// Error writes the provided error to the response. This method may only be invoked once.
+	Error(err error)
 }
 
 // Connecter is a storage object that responds to a connection request
 type Connecter interface {
-	// Connect returns a ConnectHandler that will handle the request/response for a request
-	Connect(ctx api.Context, id string, options runtime.Object) (ConnectHandler, error)
+	// Connect returns an http.Handler that will handle the request/response for a given API invocation.
+	// The provided responder may be used for common API responses.
+	Connect(ctx api.Context, id string, options runtime.Object, r Responder) (http.Handler, error)
```

Comment:
The caller of the handler has access to the same body - the responder
contract is "invoke Responder and return", while all other flows may
involve custom output.  The primary goal was avoiding a) the handler having
to know any of the details the api server abstracts (the version, the
content type, etc, how errors are written), b) making the http.Handler
terminal (no subsequent logic)

On Mon, Oct 12, 2015 at 5:07 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/api/rest/rest.go
> <https://github.com/kubernetes/kubernetes/pull/15442#discussion_r41801366>
> :
>
> >  }
> >
> >  // Connecter is a storage object that responds to a connection request
> >  type Connecter interface {
> > -	// Connect returns a ConnectHandler that will handle the request/response for a request
> > -	Connect(ctx api.Context, id string, options runtime.Object) (ConnectHandler, error)
> > +	// Connect returns an http.Handler that will handle the request/response for a given API invocation.
> > +	// The provided responder may be used for common API responses.
> > +	Connect(ctx api.Context, id string, options runtime.Object, r Responder) (http.Handler, error)
>
> I guess specifically what I'm confused about is that the handler will get
> a resp.Body to write to, but it has no way to plumb that response.Body into
> the Responder-- I'm sure I just don't understand the desired flow of
> information, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/15442/files#r41801366>.
>